tests/guest-tools/win/__init__.py

```diff
@@ -0,0 +1,123 @@
+import base64
+import logging
+import re
+from typing import List
+from lib.common import wait_for
+from lib.vm import VM
+
+
+def try_get_and_store_ip_serial(vm: VM, timeout: int):
+    domid = vm.param_get("dom-id")
+    logging.debug(f"Domain ID {domid}")
+    command = [
+        "xl",
+        "console",
+        "-t",
+        "serial",
+        domid,
+        "|",
+        "grep",
+        r"~xcp-ng-tests~.*~end~",
+        "|",
+        "head",
+        "-n",
+        "1",
+    ]
+    if timeout > 0:
+        command = ["timeout", str(timeout)] + command
+    report = vm.host.ssh(command)
+    logging.debug(f"Got report: {report}")
+    match = re.match("~xcp-ng-tests~(.*)=(.*)~end~", report)
+    if not match:
+        raise Exception(f"invalid IP report {report}")
+    return match[2]
+
+
+def encode_ps_command(cmd: str):
+    return base64.b64encode(cmd.encode("utf-16-le")).decode("ascii")
+
+
+def is_drivers_installed(vm: VM):
+    cmd = encode_ps_command(
+        r"""
+        $ProgressPreference = 'SilentlyContinue';
+        Get-PnpDevice -PresentOnly |
+        Where-Object CompatibleID -icontains 'PCI\VEN_5853' |
+        Select-Object -ExpandProperty Problem
+        """
+    )
+    output = vm.ssh(
+        ["powershell.exe", "-noprofile", "-noninteractive", "-encodedcommand", cmd]
+    )
+    logging.debug(f"Installed Xen device status: {output}")
+    if output == "CM_PROB_NONE":
+        return True
+    elif output == "CM_PROB_FAILED_INSTALL":
+        return False
+    else:
+        raise Exception(f"Unknown problem status {output}")
+
+
+def run_command_powershell(vm: VM, program: str, args: List[str]):
+    """
+    Run command under powershell to retrieve exit codes higher than 255.
+
+    Backslash-safe.
+    """
+    args = " ".join(args)
+    # ProgressPreference is needed to suppress any cli\xml progress output
+    # as it's not filtered away from stdout by default, and we're grabbing stdout
+    cmd = encode_ps_command(
+        f"""
+        $ProgressPreference = 'SilentlyContinue';
+        Write-Output (Start-Process -Wait -PassThru {program} -ArgumentList '{args}').ExitCode
+        """
+    )
+    return int(
+        vm.ssh(
+            [
+                "powershell.exe",
+                # suppress the "Preparing modules for first use" output
+                "-noprofile",
+                "-noninteractive",
+                "-encodedcommand",
+                cmd,
+            ]
+        )
+    )
+
+
+def start_background_powershell(vm: VM, cmd: str):
```

Comment:
The main motivation was to retrieve exit codes higher than 255 (which ssh didn't like) and to avoid Git Bash's path escaping behavior, which undesirably transforms arguments (e.g. `/x` -> `x:/`)